File: rl/physics.py
"""
Direct Python port of js/physics.js — identical maths, no DOM, no canvas.

All constants match the browser game so the RL agent trains on the same
physics the player experiences.
"""

import math

# ── constants (mirror of PHYSICS object in physics.js) ──────────────────────
TWO_PI               = 2 * math.pi
NO_GO_HALF_ANGLE     = math.radians(40)   # dead zone either side of true wind
DRAG                 = 0.055              # quadratic drag; terminal ≈ sqrt(MAX_DRIVE/DRAG)
MAX_DRIVE            = 0.35              # peak force at beam reach (~90° TWA)
TURN_RATE            = 3.0              # rad/s at full rudder
TACK_SPEED_FACTOR    = 0.40             # speed fraction kept on a tack
TACK_PENALTY_DURATION = 2.5            # seconds of reduced drive post-tack
LEEWAY_COEFF         = 0.05             # cosmetic sideways slip
MASS                 = 1.0
PIXELS_PER_UNIT      = 50              # canvas px per speed-unit per second

# Wind comes FROM the top of the canvas (90° in math coords = "up").
TRUE_WIND_DIR = math.radians(90)

# ── state ────────────────────────────────────────────────────────────────────

def get_initial_state():
    return {
        'x':                  400.0,
        'y':                  700.0,    # below start line
        'vx':                 0.0,
        'vy':                 0.0,
        'heading':            math.radians(40),  # ~50° TWA starboard — drives immediately
        'speed':              0.0,
        'sheet_pct':          0.80,
        'rudder':             0.0,
        'tack_penalty_timer': 0.0,
        'last_twa':           None,
        'tacking':            False,
    }

# ── core physics step ─────────────────────────────────────────────────────────

def step(state, controls, dt):
    """
    Pure function: (state, controls, dt) -> new_state.
    controls = {'rudder': -1..1, 'sheet_delta': float}
    """
    s = dict(state)   # shallow copy — all values are scalars

    # Apply controls
    s['rudder']    = max(-1.0, min(1.0, controls.get('rudder', 0.0)))
    s['sheet_pct'] = max(0.0,  min(1.0, s['sheet_pct'] + controls.get('sheet_delta', 0.0) * dt))

    twa     = signed_twa(s['heading'])
    abs_twa = abs(twa)

    # ── Tack detection (sign flip of TWA) ────────────────────────────────────
    last = s['last_twa']
    if last is not None and last != 0 and math.copysign(1, twa) != math.copysign(1, last):
        s['speed']             *= TACK_SPEED_FACTOR
        s['tack_penalty_timer'] = TACK_PENALTY_DURATION
        s['tacking']            = True
    if s['tack_penalty_timer'] <= 0:
        s['tacking'] = False
    s['last_twa']           = twa
    s['tack_penalty_timer'] = max(0.0, s['tack_penalty_timer'] - dt)

    # ── Drive force ───────────────────────────────────────────────────────────
    drive = 0.0
    if abs_twa > NO_GO_HALF_ANGLE:
        drive = _drive_from_twa(abs_twa, s['sheet_pct'])
        if s['tack_penalty_timer'] > 0:
            ramp   = 1.0 - s['tack_penalty_timer'] / TACK_PENALTY_DURATION
            drive *= ramp * ramp

    # ── Speed (net force = drive − drag) ─────────────────────────────────────
    drag       = DRAG * s['speed'] ** 2
    s['speed'] = max(0.0, s['speed'] + (drive - drag) / MASS * dt)

    # ── Position ──────────────────────────────────────────────────────────────
    leeway_angle = (
        math.copysign(LEEWAY_COEFF * s['speed'], twa)
        if abs_twa < math.radians(100) else 0.0
    )
    move_angle = s['heading'] + leeway_angle
    s['vx']    = math.cos(move_angle) * s['speed']
    s['vy']    = math.sin(move_angle) * s['speed']

    s['x'] +=  s['vx'] * dt * PIXELS_PER_UNIT
    s['y'] -= s['vy'] * dt * PIXELS_PER_UNIT   # canvas y is inverted

    # ── Heading ───────────────────────────────────────────────────────────────
    s['heading']  = (s['heading'] + s['rudder'] * TURN_RATE * dt) % TWO_PI

    return s

# ── helpers ───────────────────────────────────────────────────────────────────

def signed_twa(heading):
    """
    Signed True Wind Angle. Positive = starboard tack, negative = port tack.
    Range (-π, π).
    """
    twa = TRUE_WIND_DIR - heading
    # Normalise to (-π, π) — equivalent to JS while-loops
    return (twa + math.pi) % TWO_PI - math.pi


def _drive_from_twa(abs_twa, sheet_pct):
    """
    Drive force as a function of absolute TWA and sheet trim.
    Peaks at ~90° TWA (beam reach). Creates a real VMG tradeoff upwind.
    """
    ng   = NO_GO_HALF_ANGLE
    beam = math.pi / 2

    if abs_twa <= beam:
        # Quarter-sine ramp: 0 at no-go edge → 1.0 at beam reach
        t         = (abs_twa - ng) / (beam - ng)
        raw_drive = math.sin(t * math.pi / 2)
    else:
        # Gentle decay: beam reach → dead run
        t         = (abs_twa - beam) / (math.pi - beam)
        raw_drive = 1.0 - 0.4 * t

    # Optimal sheet: tight close-hauled (0.90) → well eased on a run (→0.25)
    optimal_sheet    = max(0.15, 0.90 - 0.65 * ((abs_twa - ng) / (math.pi - ng)))
    trim_error       = abs(sheet_pct - optimal_sheet)
    trim_efficiency  = max(0.15, 1.0 - 3.0 * trim_error ** 2)

    return MAX_DRIVE * raw_drive * trim_efficiency


def get_trim_status(state):
    """Returns 'EASE SAIL', 'TRIM OK', or 'OVERSHEET'."""
    abs_twa      = abs(signed_twa(state['heading']))
    optimal      = max(0.15, 0.90 - 0.65 * ((abs_twa - NO_GO_HALF_ANGLE) / (math.pi - NO_GO_HALF_ANGLE)))
    diff         = state['sheet_pct'] - optimal
    if diff >  0.12: return 'OVERSHEET'
    if diff < -0.12: return 'EASE SAIL'
    return 'TRIM OK'


def get_vmg(state):
    """Velocity Made Good toward the upwind mark (positive = making progress)."""
    twa = signed_twa(state['heading'])
    return state['speed'] * math.cos(abs(twa))

File: js/physics.js
// Sailing physics — pure JS, no DOM, no canvas.
// Coordinate system: x = right, y = up (math). Canvas y is inverted in renderer.
// Wind comes FROM the top of the canvas (TRUE_WIND_DIR = 90°, blows downward).
// Boat heading 90° = bow pointing straight into the wind (the no-go zone).

const TWO_PI = Math.PI * 2;
const DEG = Math.PI / 180;

const PHYSICS = {
  // No-go half-angle: zero drive within this of the wind direction
  NO_GO_HALF_ANGLE: 40 * DEG,

  // Drag (quadratic). Terminal speed ≈ sqrt(MAX_DRIVE / DRAG)
  DRAG: 0.016,

  // Peak drive force (at ~90° TWA = beam reach, best trim)
  MAX_DRIVE: 0.32,

  // Turning rate radians/s at full rudder (scales with a speed factor)
  TURN_RATE: 1.2,

  // Tack penalty
  TACK_SPEED_FACTOR: 0.38,
  TACK_PENALTY_DURATION: 2.8,

  // Cosmetic leeway: tiny sideways slip when beating
  LEEWAY_COEFF: 0.05,

  MASS: 1.0,

  // Canvas pixels per speed-unit per second (keeps movement frame-rate independent)
  PIXELS_PER_UNIT: 65,
};

// Wind FROM 90° (top of canvas). Blows toward 270° (downward on screen).
const TRUE_WIND_DIR = 90 * DEG;

export function getInitialState() {
  return {
    x: 400,
    y: 700,             // below start line — gives a few seconds to reach it
    vx: 0,
    vy: 0,
    // 40° heading → ~50° TWA → starboard tack, solidly outside no-go, driving immediately
    heading: 40 * DEG,
    speed: 0,
    sheetPct: 0.80,     // reasonable default for close-hauled
    rudder: 0,
    tackPenaltyTimer: 0,
    lastTWA: null,
    tacking: false,
  };
}

// Pure step function — no side effects, no DOM.
// controls = { rudder: -1..1, sheetDelta: number (fraction/s) }
export function step(state, controls, dt) {
  const s = { ...state };

  // Apply controls
  s.rudder   = Math.max(-1, Math.min(1, controls.rudder  ?? 0));
  s.sheetPct = Math.max(0,  Math.min(1, s.sheetPct + (controls.sheetDelta ?? 0) * dt));

  // True Wind Angle: + = starboard tack, - = port tack
  const twa    = signedTWA(s.heading);
  const absTWA = Math.abs(twa);

  // Tack detection (sign flip of TWA)
  if (s.lastTWA !== null && s.lastTWA !== 0 && Math.sign(twa) !== Math.sign(s.lastTWA)) {
    s.speed *= PHYSICS.TACK_SPEED_FACTOR;
    s.tackPenaltyTimer = PHYSICS.TACK_PENALTY_DURATION;
    s.tacking = true;
  }
  if (s.tackPenaltyTimer <= 0) s.tacking = false;
  s.lastTWA = twa;
  s.tackPenaltyTimer = Math.max(0, s.tackPenaltyTimer - dt);

  // Drive force
  let drive = 0;
  if (absTWA > PHYSICS.NO_GO_HALF_ANGLE) {
    drive = driveFromTWA(absTWA, s.sheetPct);
    if (s.tackPenaltyTimer > 0) {
      // Ramp drive back in quadratically after a tack
      const ramp = 1 - s.tackPenaltyTimer / PHYSICS.TACK_PENALTY_DURATION;
      drive *= ramp * ramp;
    }
  }

  // Speed update (drag is quadratic)
  const drag = PHYSICS.DRAG * s.speed * s.speed;
  s.speed = Math.max(0, s.speed + ((drive - drag) / PHYSICS.MASS) * dt);

  // Leeway: small lateral slip only when close-hauled
  const leewayAngle = (absTWA < 100 * DEG)
    ? Math.sign(twa) * PHYSICS.LEEWAY_COEFF * s.speed
    : 0;

  const moveAngle = s.heading + leewayAngle;
  s.vx = Math.cos(moveAngle) * s.speed;
  s.vy = Math.sin(moveAngle) * s.speed;

  // Position (frame-rate independent)
  s.x +=  s.vx * dt * PHYSICS.PIXELS_PER_UNIT;
  s.y -= s.vy * dt * PHYSICS.PIXELS_PER_UNIT;  // canvas y inverted

  // Steering: small minimum effect at zero speed so the player isn't stranded
  const steerEffect = 0.25 + 0.75 * Math.min(s.speed / 2.0, 1.0);
  s.heading += s.rudder * PHYSICS.TURN_RATE * steerEffect * dt;
  s.heading = ((s.heading % TWO_PI) + TWO_PI) % TWO_PI;

  return s;
}

// Signed TWA. Range (-π, π). Positive = starboard tack.
function signedTWA(heading) {
  let twa = TRUE_WIND_DIR - heading;
  while (twa >  Math.PI) twa -= TWO_PI;
  while (twa < -Math.PI) twa += TWO_PI;
  return twa;
}

// Drive force as a function of TWA and sheet trim.
//
// Drive peaks at ~90° TWA (beam reach). This creates a real VMG tradeoff:
//   - Close-hauled (~45° TWA): slow speed but good angle → moderate VMG
//   - Beam reach (90° TWA):    fast speed but perpendicular → zero VMG upwind
//   - Optimal VMG groove:      ~50–60° TWA (player has to find it)
//
function driveFromTWA(absTWA, sheetPct) {
  const ng   = PHYSICS.NO_GO_HALF_ANGLE;  // 40° — no-go edge
  const beam = Math.PI / 2;               // 90° — peak drive

  let rawDrive;
  if (absTWA <= beam) {
    // Quarter-sine ramp: 0 at no-go edge, 1.0 at beam reach
    const t = (absTWA - ng) / (beam - ng);
    rawDrive = Math.sin(t * Math.PI / 2);
  } else {
    // Gentle decay from beam reach to dead run
    const t = (absTWA - beam) / (Math.PI - beam);
    rawDrive = 1.0 - 0.4 * t;
  }

  // Optimal sheet: tight close-hauled (0.90), eased reaching/running (→0.25)
  const optimalSheet = Math.max(0.15, 0.90 - 0.65 * ((absTWA - ng) / (Math.PI - ng)));
  const trimError    = Math.abs(sheetPct - optimalSheet);
  const trimEfficiency = Math.max(0.15, 1.0 - 3.0 * trimError * trimError);

  return PHYSICS.MAX_DRIVE * rawDrive * trimEfficiency;
}

// For trim feedback on the HUD
export function getTrimStatus(state) {
  const absTWA = Math.abs(signedTWA(state.heading));
  const ng = PHYSICS.NO_GO_HALF_ANGLE;
  const optimalSheet = Math.max(0.15, 0.90 - 0.65 * ((absTWA - ng) / (Math.PI - ng)));
  const diff = state.sheetPct - optimalSheet;
  if (diff >  0.12) return 'OVERSHEET';
  if (diff < -0.12) return 'EASE SAIL';
  return 'TRIM OK';
}

// Flat observation vector for RL
export function getObservation(state, raceState) {
  const twa = signedTWA(state.heading);
  return [
    state.x / 800,
    state.y / 800,
    Math.cos(state.heading),
    Math.sin(state.heading),
    twa / Math.PI,
    state.speed / 5,
    state.sheetPct,
    state.tackPenaltyTimer / PHYSICS.TACK_PENALTY_DURATION,
    raceState?.timeToStart  ?? 0,
    raceState?.distToFinish ?? 0,
  ];
}

export { signedTWA, TRUE_WIND_DIR, PHYSICS };
